src/fuzz/fuzz-udev-rules.c

```diff
@@ -1,106 +1,36 @@
 /* SPDX-License-Identifier: LGPL-2.1+ */
 
-#include <errno.h>
-#include <sched.h>
-#include <sys/mount.h>
-#include <unistd.h>
+#include <stdio.h>
 
 #include "fd-util.h"
 #include "fs-util.h"
 #include "fuzz.h"
-#include "log.h"
-#include "mkdir.h"
-#include "rm-rf.h"
-#include "string-util.h"
 #include "tests.h"
+#include "tmpfile-util.h"
 #include "udev-rules.h"
 
-static struct fakefs {
-        const char *target;
-        bool ignore_mount_error;
-        bool is_mounted;
-} fakefss[] = {
-        { "/sys",                    false, false },
-        { "/dev",                    false, false },
-        { "/run",                    false, false },
-        { "/etc",                    false, false },
-        { UDEVLIBEXECDIR "/rules.d", true, false },
-};
-
-static int setup_mount_namespace(void) {
-        static thread_local bool is_namespaced = false;
-
-        if (is_namespaced)
-                return 1;
-
-        if (unshare(CLONE_NEWNS) < 0)
-                return log_error_errno(errno, "Failed to call unshare(): %m");
-
-        if (mount(NULL, "/", NULL, MS_SLAVE|MS_REC, NULL) < 0)
-                return log_error_errno(errno, "Failed to mount / as private: %m");
-
-        is_namespaced = true;
-
-        return 1;
-}
-
-static int setup_fake_filesystems(const char *runtime_dir) {
-        for (unsigned i = 0; i < ELEMENTSOF(fakefss); i++) {
-                if (mount(runtime_dir, fakefss[i].target, NULL, MS_BIND, NULL) < 0) {
-                        log_full_errno(fakefss[i].ignore_mount_error ? LOG_DEBUG : LOG_ERR, errno, "Failed to mount %s: %m", fakefss[i].target);
-                        if (!fakefss[i].ignore_mount_error)
-                                return -errno;
-                } else
-                        fakefss[i].is_mounted = true;
-        }
-
-        return 0;
-}
-
-static int cleanup_fake_filesystems(const char *runtime_dir) {
-        for (unsigned i = 0; i < ELEMENTSOF(fakefss); i++) {
-                if (!fakefss[i].is_mounted)
-                        continue;
-
-                if (umount(fakefss[i].target) < 0) {
-                        log_full_errno(fakefss[i].ignore_mount_error ? LOG_DEBUG : LOG_ERR, errno, "Failed to umount %s: %m", fakefss[i].target);
-                        if (!fakefss[i].ignore_mount_error)
-                                return -errno;
-                } else
-                        fakefss[i].is_mounted = false;
-        }
-        return 0;
-}
-
 int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
         _cleanup_(udev_rules_freep) UdevRules *rules = NULL;
-        _cleanup_(rm_rf_physical_and_freep) char *runtime_dir = NULL;
-        FILE *f = NULL;
-
-        (void) setup_mount_namespace();
-
-        assert_se(runtime_dir = setup_fake_runtime_dir());
-
-        if (setup_fake_filesystems(runtime_dir) < 0) {
```

Comment:
In its current form, the fuzz target is kind of half useful because it just parses rules without applying them, which causes it to miss issues like https://github.com/systemd/systemd/pull/15064. As far as I can remember the point of those fake filesystems was to make it easier to call `udev_event_execute_rules` and `udev_event_execute_run` without destroying anything.

Regarding https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=20142, I think it's a bug somewhere in the OSS-Fuzz sandbox and should be fixed there regardless of whether `fuzz-udev-rules` is run as `root` or not.

Reply:
I think the fuzz target *is* useful for checking the parser code. With the patch the check is effectively the same, just faster and with less setup. A different fuzz target for rule execution wouldn't be bad, but if somebody writes it, think it should become a new independent executable. Anyway, the code is in git, so it's trivial to resurrect it later if desired.

Reply:
> With the patch the check is effectively the same, just faster and with less setup

Agreed.

> if somebody writes it, think it should become a new independent executable

I think since `fuzz-udev-rules` already parses rules it also should apply them if `udev_rules_parse_file` doesn't complain. Anyway, since at this point it seems to be a theoretical scenario I don't think it matters much.

> Anyway, the code is in git, so it's trivial to resurrect it later if desired.

Can't argue with that :-)